tls/design considerations.md

```diff
@@ -0,0 +1,47 @@
+# Design considerations for the libp2p TLS Handshake
+
+## Requirements 
+
+There are two main requirements that prevent us from using the straightforward way to run a TLS handshake (which would be to simply use the host key to create a self-signed certificate).
+
+1. We want to use different key types: RSA, ECDSA, and Ed25519, Secp256k1 (and maybe more in the future?).
+2. We want to be able to send the key type along with the key (see https://github.com/libp2p/specs/issues/111).
+
+The first point is problematic in practice, because Go currently only supports RSA and ECDSA certificates. Support for Ed25519 was planned for Go 1.12, but was deferred recently, and the Go team is now evaluating interest in this in order to prioritze their work, so this might or might not happen in Go 1.13. I'm not aware of any plans for Secp256k1 at the moment.
+The second requirement implies that we might want add some additional (free-form) information to the handshake, and we need to find a field to stuff that into.
+
+The handshake protocol described here:
+* supports arbitrary keys, independent from what the signature algorithms implemented by the TLS library used
+* defines a way how future versions of this protocol might be negotiated without requiring any out-of-band information and additional roundtrips
+
+
+## Design Choices
+
+### TLS 1.3 - What about older versions?
+
+The handshake protocol requires TLS 1.3 support. This means that the handshake between to peers that have never communicated before will typically complete in just a single roundtrip. With older TLS versions, a handshake typically takes two roundtrips. By not specifying support for older TLS versions, we increase perfomance and simplify the protocol.
+
+
+### Why we're not using the host key for the certificate
+
+The current proposal uses a self-signed certificate to carry the host's public key in the libp2p Public Key Extension. The key used to generate the self-signed certificate has no relationship with the host key. This key can be generated for every single connection, or can be generated at boot time.
```

Comment:
IMO another reason for not using the host key directly is to achieve Perfect Forward Secrecy.

Reply:
TLS 1.3 uses ephemeral Diffie-Hellman for the key exchange mechanism, so it's always PFS, no matter what kind of certificate you use.